lib/gl/renderer_core.cpp

```diff
@@ -0,0 +1,795 @@
+            && verts[t_i*3+2].clipCoord >= 0.f)
+        {
+            // triangle fully in the unclipped region
+            for (int vert_i = 0; vert_i < 3; vert_i++) {
+                cbuf.triangles.emplace_back(
+                        XFBPostTransform(verts[t_i*3 + vert_i], half_w, half_h)); 
+            }
+        }
+        else if (verts[3*t_i].clipCoord < 0.f
+            && verts[3*t_i+1].clipCoord < 0.f
+            && verts[3*t_i+2].clipCoord < 0.f)
+        {
+            //triangle fully in clipped region
+            continue;
+        }
+        else
+        {
+            //clip through middle of triangle
+            for (int vert_i = 0; vert_i < 3; vert_i++) {
+                int i_a = 3*t_i+vert_i;
+                int i_b = 3*t_i+((vert_i+1) % 3);
+                int i_c = 3*t_i+((vert_i+2) % 3);
+                // find two points on the same side of clip plane
+                if ((verts[i_a].clipCoord < 0.f) == (verts[i_b].clipCoord < 0.f)) {
+                    //pts a, b are on same side of clip plane, c on other side
+                    //perspective-correct interpolation factors, needed for colors
+                    float c_w_a = verts[i_a].clipCoord / verts[i_a].pos[3];
+                    float c_w_b = verts[i_b].clipCoord / verts[i_b].pos[3];
+                    float c_w_c = verts[i_c].clipCoord / verts[i_c].pos[3];
+                    //compute clip pts
+                    glm::vec4 pos[2], color[2];
+                    // a --- n_0 --- c
+                    pos[0] = (glm::make_vec4(verts[i_a].pos) * verts[i_c].clipCoord
+                              - glm::make_vec4(verts[i_c].pos) * verts[i_a].clipCoord);
+                    color[0] = (glm::make_vec4(verts[i_a].color) * c_w_c
+                                - glm::make_vec4(verts[i_c].color) * c_w_a)
+                                / (c_w_c - c_w_a);
+                    // b --- n_1 --- c
+                    pos[1] = (glm::make_vec4(verts[i_b].pos) * verts[i_c].clipCoord
+                              - glm::make_vec4(verts[i_c].pos) * verts[i_b].clipCoord);
+                    color[1] = (glm::make_vec4(verts[i_b].color) * c_w_c
+                                - glm::make_vec4(verts[i_c].color) * c_w_b)
+                                / (c_w_c - c_w_b);
+                    for (int i = 0; i < 2; i++)
+                    {
+                        // perform transform to device coords
+                        pos[i] /= pos[i].w;
+                        pos[i].x *= half_w; pos[i].x += half_w;
+                        pos[i].y *= half_h; pos[i].y += half_h;
+                    }
+
+                    if (verts[i_c].clipCoord < 0.f) {
+                        //pts a, b are in clip plane
+                        //create quadrilateral a -- n_0 -- n_1 -- b
+                        cbuf.triangles.emplace_back(XFBPostTransform(verts[i_a], half_w, half_h));
+                        cbuf.triangles.emplace_back(pos[0], color[0]);
```

Comment:
This doesn't compile (vec3 vs. vec4?)

Reply:
I wasn't able to replicate this issue with gcc 5.5 on OSX. Could you post the compile error you're getting?

Reply:
Sure:
```
g++ -std=c++11 -O3 -std=c++11 -I/home/jakub/mfem -DGLVIS_MULTISAMPLE=4 -DGLVIS_MS_LINEWIDTH=1.4 -DGLVIS_OGL3 -I/usr/include/freetype2 -I/usr/include -I/usr/include -I/usr/include -I/usr/include -DSDL_SCREENSHOTS -o lib/gl/renderer_core.o -c lib/gl/renderer_core.cpp
In file included from /usr/include/x86_64-linux-gnu/c++/5/bits/c++allocator.h:33:0,
                 from /usr/include/c++/5/bits/allocator.h:46,
                 from /usr/include/c++/5/vector:61,
                 from lib/gl/types.hpp:14,
                 from lib/gl/attr_traits.hpp:1,
                 from lib/gl/renderer_core.cpp:1:
/usr/include/c++/5/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = gl3::FeedbackVertex; _Args = {glm::tvec4<float, (glm::precision)0u>&, glm::tvec4<float, (glm::precision)0u>&}; _Tp = gl3::FeedbackVertex]’:
/usr/include/c++/5/bits/alloc_traits.h:530:4:   required from ‘static void std::allocator_traits<std::allocator<_Tp1> >::construct(std::allocator_traits<std::allocator<_Tp1> >::allocator_type&, _Up*, _Args&& ...) [with _Up = gl3::FeedbackVertex; _Args = {glm::tvec4<float, (glm::precision)0u>&, glm::tvec4<float, (glm::precision)0u>&}; _Tp = gl3::FeedbackVertex; std::allocator_traits<std::allocator<_Tp1> >::allocator_type = std::allocator<gl3::FeedbackVertex>]’
/usr/include/c++/5/bits/vector.tcc:96:30:   required from ‘void std::vector<_Tp, _Alloc>::emplace_back(_Args&& ...) [with _Args = {glm::tvec4<float, (glm::precision)0u>&, glm::tvec4<float, (glm::precision)0u>&}; _Tp = gl3::FeedbackVertex; _Alloc = std::allocator<gl3::FeedbackVertex>]’
lib/gl/renderer_core.cpp:697:69:   required from here
/usr/include/c++/5/ext/new_allocator.h:120:4: error: no matching function for call to ‘gl3::FeedbackVertex::FeedbackVertex(glm::tvec4<float, (glm::precision)0u>&, glm::tvec4<float, (glm::precision)0u>&)’
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^
In file included from lib/gl/renderer_core.hpp:3:0,
                 from lib/gl/attr_traits.hpp:2,
                 from lib/gl/renderer_core.cpp:1:
lib/gl/renderer.hpp:58:5: note: candidate: gl3::FeedbackVertex::FeedbackVertex(glm::vec3, glm::vec4)
     FeedbackVertex(glm::vec3 pos, glm::vec4 c)
     ^
lib/gl/renderer.hpp:58:5: note:   no known conversion for argument 1 from ‘glm::tvec4<float, (glm::precision)0u>’ to ‘glm::vec3 {aka glm::tvec3<float, (glm::precision)0u>}’
lib/gl/renderer.hpp:57:5: note: candidate: gl3::FeedbackVertex::FeedbackVertex()
     FeedbackVertex() = default;
     ^
lib/gl/renderer.hpp:57:5: note:   candidate expects 0 arguments, 2 provided
lib/gl/renderer.hpp:52:8: note: candidate: constexpr gl3::FeedbackVertex::FeedbackVertex(const gl3::FeedbackVertex&)
 struct FeedbackVertex
        ^
lib/gl/renderer.hpp:52:8: note:   candidate expects 1 argument, 2 provided
lib/gl/renderer.hpp:52:8: note: candidate: constexpr gl3::FeedbackVertex::FeedbackVertex(gl3::FeedbackVertex&&)
lib/gl/renderer.hpp:52:8: note:   candidate expects 1 argument, 2 provided
makefile:244: recipe for target 'lib/gl/renderer_core.o' failed
```

Reply:
I recall that some versions of GLM suck, maybe it's that?
```
$ dpkg -l '*libglm*'
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                                    Version                  Architecture             Description
+++-=======================================-========================-========================-===================================================================================
ii  libglm-dev                              0.9.7.2-1                all                      C++ library for OpenGL GLSL type-based mathematics
```

Reply:
Hmm, that might be the issue; 0.9.7.2 was released in January 2016.

GLM is header-only, so you should be able to download a more recent version and point GLM_DIR in your make command to the new version of GLM.

Reply:
It is old, but it's part of Ubuntu 16.04 LTS, which is supported until 2021 and some folks may still be using it. Also the fix seems to be simple actually, just add another constructor in `FeedbackVertex`:
```
    FeedbackVertex(glm::vec4 pos, glm::vec4 c)
        : position(pos), color(c) { }
```